packages/ckeditor5-markdown-gfm/src/pastefrommarkdownexperimental.ts

```diff
@@ -0,0 +1,166 @@
+		const viewDocument = view.document;
+
+		const clipboardPipeline: ClipboardPipeline = editor.plugins.get( 'ClipboardPipeline' );
+
+		let shiftPressed = false;
+
+		this.listenTo<ViewDocumentKeyDownEvent>( viewDocument, 'keydown', ( evt, data ) => {
+			shiftPressed = data.shiftKey;
+		} );
+
+		this.listenTo<ClipboardInputTransformationEvent>( clipboardPipeline, 'inputTransformation', ( evt, data ) => {
+			if ( shiftPressed ) {
+				return;
+			}
+
+			const dataAsTextHtml = data.dataTransfer.getData( 'text/html' );
+
+			if ( !dataAsTextHtml ) {
+				const dataAsTextPlain = data.dataTransfer.getData( 'text/plain' );
+
+				data.content = this._gfmDataProcessor.toView( dataAsTextPlain );
+
+				return;
+			}
+
+			const markdownFromHtml = this._parseMarkdownFromHtml( dataAsTextHtml );
+
+			if ( markdownFromHtml ) {
+				data.content = this._gfmDataProcessor.toView( markdownFromHtml );
+			}
+		} );
+	}
+
+	/**
+	 * Determines if code copied from a website in `text/html` type can be parsed as markdown.
+	 * It removes any OS specific HTML tags e.g. <meta> on Mac OS and <!--StartFragment--> on Windows.
+	 * Then removes a single wrapper HTML tag or wrappers for sibling tags, and if there are no more tags left,
+	 * returns the remaining text. Returns null, if there are any remaining HTML tags detected.
+	 *
+	 * @param htmlString Clipboard content in `text/html` type format.
+	 */
+	private _parseMarkdownFromHtml( htmlString: string ): string | null {
+		const withoutOsSpecificTags = this._removeOsSpecificTags( htmlString );
+		const withoutWrapperTag = this._removeFirstLevelWrapperTagsAndBrs( withoutOsSpecificTags );
+
+		if ( this._containsAnyRemainingHtmlTags( withoutWrapperTag ) ) {
+			return null;
+		}
+
+		return this._replaceHtmlReservedEntitiesWithCharacters( withoutWrapperTag );
+	}
+
+	/**
+	 * Removes OS specific tags.
+	 *
+	 * @param htmlString Clipboard content in `text/html` type format.
+	 */
+	private _removeOsSpecificTags( htmlString: string ): string {
+		// Removing <meta> tag present on Mac.
+		const withoutMetaTag = htmlString.replace( /^<meta\b[^>]*>/, '' ).trim();
+		// Removing <html> tag present on Windows.
+		const withoutHtmlTag = withoutMetaTag.replace( /^<html>/, '' ).replace( /<\/html>$/, '' ).trim();
+		// Removing <body> tag present on Windows.
+		const withoutBodyTag = withoutHtmlTag.replace( /^<body>/, '' ).replace( /<\/body>$/, '' ).trim();
+
+		// Removing <!--StartFragment--> tag present on Windows.
+		return withoutBodyTag.replace( /^<!--StartFragment-->/, '' ).replace( /<!--EndFragment-->$/, '' ).trim();
+	}
+
+	/**
+	 * Removes multiple HTML wrapper tags from a list of sibling HTML tags.
+	 *
+	 * @param htmlString Clipboard content without any OS specific tags.
+	 */
+	private _removeFirstLevelWrapperTagsAndBrs( htmlString: string ): string {
+		const tempDiv = document.createElement( 'div' );
+		tempDiv.innerHTML = htmlString;
+
+		const outerElements = tempDiv.querySelectorAll( ':not(:empty)' );
```

Comment:
This selector is contrary to the method name. It collects all the non-empty elements, not only the 'first-level'. This looks like it would unwrap all inline tags also. I'm not sure if this is the desired behavior.

Reply:
Changed to 
```js
const outerElements = tempElement.querySelectorAll( ':scope > *' );
```